File: tamm_parser/semant.cc
#include "semant.h"

void check_ast(TranslationUnit *root, SymbolTable &symtab) {
    CompoundElemList *celist = root->celist;
    while (celist != nullptr) {
        check_CompoundElem(celist->head, symtab);
        celist = celist->tail;
    }
}

void check_CompoundElem(CompoundElem *celem, SymbolTable &symtab) {
    ElemList *elist = celem->elist;
    while (elist != nullptr) {
        check_Elem(elist->head, symtab);
        elist = elist->tail;
    }
}

void check_Elem(Elem *elem, SymbolTable &symtab) {
    Elem *e = elem;
    if (e == nullptr) return;

    switch (e->kind) {
        case Elem::is_DeclList:
            check_DeclList(elem->u.d, symtab);
            break;
        case Elem::is_Statement:
            check_Stmt(e->u.s, symtab);
            break;
        default:
            std::cerr << "Not a Declaration or Statement!\n";
            std::exit(EXIT_FAILURE);
    }
}

void check_DeclList(DeclList *decllist, SymbolTable &symtab) {
    DeclList *dl = decllist;
    while (dl != nullptr) {
        check_Decl(dl->head, symtab);
        dl = dl->tail;
    }
}

void verifyVarDecl(SymbolTable &symtab, tamm_string name, int line_no) {
    if (symtab.find(name) != symtab.end()) {
        std::cerr << "Error at line " << line_no << ": " << name << " is already defined\n";
        std::exit(EXIT_FAILURE);
    }
}

void verifyRangeRef(SymbolTable &symtab, tamm_string name, int line_no) {
//    if (symtab.find(name) == symtab.end()){
//        std::cerr << "Error at line " << line_no << ":range variable " << name << " is not defined\n";
//        std::exit(EXIT_FAILURE);
//    }
    const int rno = 3;
    tamm_string_array ranges = {"O", "V", "N"};
    if (!exists_index(ranges, name)) {
        std::cerr << "Error at line " << line_no << ": range " << name << " is not supported. " <<
                  "Can only be one of " << combine_indices(ranges) << std::endl;
        std::exit(EXIT_FAILURE);
    }
}

void check_Decl(Decl *d, SymbolTable &symtab) {
    switch (d->kind) {
        case Decl::is_RangeDecl: {
            verifyVarDecl(symtab, d->u.RangeDecl.name, d->lineno);
            if (d->u.RangeDecl.value <= 0) {
                std::cerr << "Error at line " << d->lineno << ": " << d->u.RangeDecl.value
                          << " is not a positive integer\n";
                std::exit(EXIT_FAILURE);
            }
            symtab.insert(SymbolTable::value_type(
                    std::string(d->u.RangeDecl.name), constcharToChar((std::to_string(d->u.RangeDecl.value)).c_str())));
        }
            break;
        case Decl::is_IndexDecl: {
            verifyVarDecl(symtab, d->u.IndexDecl.name, d->lineno);
            verifyRangeRef(symtab, d->u.IndexDecl.rangeID, d->lineno);
            symtab.insert(SymbolTable::value_type(std::string(d->u.IndexDecl.name), (d->u.IndexDecl.rangeID)));
        }
            break;
        case Decl::is_ArrayDecl: {
            tamm_string_array up_ind(d->u.ArrayDecl.ulen);
            for (int i = 0; i < d->u.ArrayDecl.ulen; i++)
                up_ind[i] = d->u.ArrayDecl.upperIndices[i];

            tamm_string_array lo_ind(d->u.ArrayDecl.llen);
            for (int i = 0; i < d->u.ArrayDecl.llen; i++)
                lo_ind[i] = d->u.ArrayDecl.lowerIndices[i];

            verifyVarDecl(symtab, d->u.ArrayDecl.name, d->lineno);
            tamm_string comb_index_list = combine_indexLists(up_ind, lo_ind);
            //std::cout << d->u.ArrayDecl.name << " -> " << comb_index_list << std::endl;
            int i = 0;
            tamm_string *ind_list = d->u.ArrayDecl.upperIndices;
            for (i = 0; i < d->u.ArrayDecl.ulen; i++) verifyRangeRef(symtab, ind_list[i], d->lineno);
            ind_list = d->u.ArrayDecl.lowerIndices;
            for (i = 0; i < d->u.ArrayDecl.llen; i++) verifyRangeRef(symtab, ind_list[i], d->lineno);

            symtab.insert(SymbolTable::value_type(std::string(d->u.ArrayDecl.name), (comb_index_list)));
        }
            break;
        default: {
            std::cerr << "Not a valid Declaration!\n";
            std::exit(EXIT_FAILURE);
        }
    }
}

void check_Stmt(Stmt *s, SymbolTable &symtab) {
    switch (s->kind) {
        case Stmt::is_AssignStmt:
            check_Exp(s->u.AssignStmt.lhs, symtab);
            //std::cout << " " << s->u.AssignStmt.astype << " "; //astype not needed since we flatten. keep it for now.
            check_Exp(s->u.AssignStmt.rhs, symtab);
            if (s->u.AssignStmt.lhs->kind != Exp::is_ArrayRef) {
                std::cerr << "Error at line " << s->u.AssignStmt.lhs->lineno
                          << ": LHS of assignment must be an array reference\n";
                std::exit(EXIT_FAILURE);
            } else if (s->u.AssignStmt.lhs->coef < 0) {
                std::cerr << "Error at line " << s->u.AssignStmt.lhs->lineno
                          << ": LHS array reference cannot be negative\n";
                std::exit(EXIT_FAILURE);
            }

//    UNCOMMENT FOR DEBUG ONLY
//    print_index_list(getIndices(s->u.AssignStmt.lhs));
//    std::cout << " = ";
//    print_index_list(getIndices(s->u.AssignStmt.rhs));
//    std::cout << "\n";
            if (!compare_index_lists(getIndices(s->u.AssignStmt.lhs), getIndices(s->u.AssignStmt.rhs))) {
                std::cerr << "Error at line " << s->u.AssignStmt.lhs->lineno
                          << ": LHS and RHS of assignment must have equal (non-summation) index sets\n";
                std::exit(EXIT_FAILURE);
            }

//    tamm_string_array lhs_aref = collectArrayRefs(s->u.AssignStmt.lhs);
//    tamm_string_array rhs_arefs = collectArrayRefs(s->u.AssignStmt.rhs);
//    if (exists_index(rhs_arefs->list,rhs_arefs->length,lhs_aref->list[0])){
//        std::cerr << "Error at line " << s->u.AssignStmt.lhs->lineno << ": array " << lhs_aref->list[0] << " cannot be assigned after being previously referenced\n";
//        std::exit(EXIT_FAILURE);
//    }
            break;
        default:
            std::cerr << "Not an Assignment Statement!\n";
            std::exit(EXIT_FAILURE);
    }
}


void check_ExpList(ExpList *expList, SymbolTable &symtab) {
    ExpList *elist = expList;
    while (elist != nullptr) {
        check_Exp(elist->head, symtab);
        elist = elist->tail;
    }
}


void verifyArrayRefName(SymbolTable &symtab, tamm_string name, int line_no) {
    if (symtab.find(name) == symtab.end()) {
        std::cerr << "Error at line " << line_no << ": array " << name << " is not defined\n";
        std::exit(EXIT_FAILURE);
    }
}

void verifyIndexRef(SymbolTable &symtab, tamm_string name, int line_no) {
    if (symtab.find(name) == symtab.end()) {
        std::cerr << "Error at line " << line_no << ": index " << name << " is not defined\n";
        std::exit(EXIT_FAILURE);
    }
}

void verifyArrayRef(SymbolTable &symtab, tamm_string name, tamm_string *inds, int len, int line_no) {
    verifyArrayRefName(symtab, name, line_no);
    int i = 0;
    for (i = 0; i < len; i++) verifyIndexRef(symtab, inds[i], line_no);
}


void check_Exp(Exp *exp, SymbolTable &symtab) {
    tamm_string_array inames;
    ExpList *el = nullptr;
    int clno = exp->lineno;
    switch (exp->kind) {
        case Exp::is_Parenth: {
            check_Exp(exp->u.Parenth.exp, symtab);
        }
            break;
        case Exp::is_NumConst: {
        }
            //std::cout << exp->u.NumConst.value << " ";
            break;
        case Exp::is_ArrayRef: {
            verifyArrayRef(symtab, exp->u.Array.name, exp->u.Array.indices, exp->u.Array.length, clno);
            inames = getIndices(exp);
            tamm_string_array all_ind1 = inames;
            tamm_string_array rnames;

            for (auto i1: all_ind1)
                rnames.push_back(symtab[i1]);

            tamm_string_array rnamesarr = rnames;
            tamm_string ulranges = symtab[exp->u.Array.name];
            tamm_string_array ulr = stringToList(ulranges);

            if (!check_array_usage(ulr, rnamesarr)) {
                std::cerr << "Error at line " << clno << ": array reference " << exp->u.Array.name << "["
                          << combine_indices(all_ind1) << "]"
                          << " must have index structure of " << exp->u.Array.name << "[" << combine_indices(ulr)
                          << "]\n";
                std::exit(EXIT_FAILURE);
            }
            //Check for repetitive indices in an array reference
            tamm_string_array uind1;

            for (auto i1: all_ind1) {
                if (!exists_index(uind1, i1))
                    uind1.push_back(i1);
            }

            tamm_string_array up_ind(exp->u.Array.length);
            for (int i = 0; i < exp->u.Array.length; i++)
                up_ind[i] = exp->u.Array.indices[i];

            for (auto i1:uind1) {
                if (count_index(all_ind1, i1) > 1) {
                    std::cerr << "Error at line " << clno << ": repetitive index " << i1 << " in array reference "
                              << exp->u.Array.name << "[" << combine_indices(up_ind) << "]\n";
                    std::exit(EXIT_FAILURE);
                }
            }
        }
            break;
        case Exp::is_Addition: {
            check_ExpList(exp->u.Addition.subexps, symtab);
            inames = getIndices(exp);
            el = exp->u.Addition.subexps;
            while (el != nullptr) {
                tamm_string_array op_inames = getIndices(el->head);
                if (!compare_index_lists(inames, op_inames)) {
                    std::cerr << "Error at line " << clno
                              << ": subexpressions of an addition must have equal index sets\n";
                    std::exit(EXIT_FAILURE);
                }

                el = el->tail;
            }
            break;
            case Exp::is_Multiplication:
                check_ExpList(exp->u.Multiplication.subexps, symtab);
            el = exp->u.Multiplication.subexps;
            tamm_string_array all_ind;

            while (el != nullptr) {
                tamm_string_array se = getIndices(el->head);
                for (auto i: se)
                    all_ind.push_back(i);
                el = el->tail;
            }

            tamm_string_array uind;
            for (auto i: all_ind) {
                if (!exists_index(uind, i))
                    uind.push_back(i);
            }

            for (auto i:uind) {
                if (count_index(all_ind, i) > 2) {
                    std::cerr << "Error at line " << clno << ": summation index " << i <<
                              " must occur exactly twice in a multiplication\n";
                    std::exit(EXIT_FAILURE);
                }
            }
        }
            break;
        default: {
            std::cerr << "Not a valid Expression!\n";
            std::exit(EXIT_FAILURE);
        }
    }
}

//get non-summation indices only
tamm_string_array getIndices(Exp *exp) {
    ExpList *el = nullptr;
    tamm_string_array p;
    switch (exp->kind) {
        case Exp::is_Parenth: {
            return getIndices(exp->u.Parenth.exp);
        }
        case Exp::is_NumConst: {
            return p;
        }
        case Exp::is_ArrayRef: {
            tamm_string_array up_ind(exp->u.Array.length);
            for (int i = 0; i < exp->u.Array.length; i++)
                up_ind[i] = strdup(exp->u.Array.indices[i]);
            return up_ind;
        }
        case Exp::is_Addition: {
            return getIndices(exp->u.Addition.subexps->head);
        }
        case Exp::is_Multiplication: {
            el = exp->u.Multiplication.subexps;
            tamm_string_array all_ind;
            while (el != nullptr) {
                tamm_string_array se = getIndices(el->head);
                for (auto i: se) {
                    all_ind.push_back(i);
                }
                el = el->tail;
            }

            tamm_string_array uind;
            for (auto i: all_ind) {
                if (count_index(all_ind, i) == 1)
                    uind.push_back(i);
            }

            tamm_string_array uniq_ind;
            for (auto i: uind) uniq_ind.push_back(strdup(i));
            return uniq_ind;
        }

        default: {
            std::cerr << "Not a valid Expression!\n";
            std::exit(EXIT_FAILURE);
        }
    }
}


void print_ExpList(ExpList *expList, tamm_string am) {
    ExpList *elist = expList;
    while (elist != nullptr) {
        print_Exp(elist->head);
        elist = elist->tail;
        if (elist != nullptr) std::cout << am << " ";
    }
}


void print_Exp(Exp *exp) {
    switch (exp->kind) {
        case Exp::is_Parenth:
            print_Exp(exp->u.Parenth.exp);
            break;
        case Exp::is_NumConst:
            std::cout << exp->u.NumConst.value << " ";
            break;
        case Exp::is_ArrayRef: {
            tamm_string_array up_ind(exp->u.Array.length);
            for (int i = 0; i < exp->u.Array.length; i++)
                up_ind[i] = exp->u.Array.indices[i];
            std::cout << exp->u.Array.name << "[" << combine_indices(up_ind) << "] ";
            break;
        }
        case Exp::is_Addition:
            print_ExpList(exp->u.Addition.subexps, "+");
            break;
        case Exp::is_Multiplication:
            print_ExpList(exp->u.Multiplication.subexps, "*");
            break;
        default:
            std::cerr << "Not a valid Expression!\n";
            std::exit(EXIT_FAILURE);
    }
}


//get all indices only once
tamm_string_array getUniqIndices(Exp *exp) {
    ExpList *el = nullptr;
    tamm_string_array p;
    switch (exp->kind) {
        case Exp::is_Parenth: {
            return getUniqIndices(exp->u.Parenth.exp);
        }

        case Exp::is_NumConst: {
            return p;
        }

        case Exp::is_ArrayRef: {
            tamm_string_array up_ind(exp->u.Array.length);
            for (int i = 0; i < exp->u.Array.length; i++)
                up_ind[i] = strdup(exp->u.Array.indices[i]);
            return up_ind;
        }

        case Exp::is_Addition: {
            return getUniqIndices(exp->u.Addition.subexps->head);
        }

        case Exp::is_Multiplication: {
            el = exp->u.Multiplication.subexps;
            tamm_string_array all_ind;
            while (el != nullptr) {
                tamm_string_array se = getUniqIndices(el->head);
                for (auto elem: se) {
                    all_ind.push_back(elem);
                }
                el = el->tail;
            }

            tamm_string_array uind;
            for (auto i: all_ind) {
                if (!exists_index(uind, i))
                    uind.push_back(i);
            }

            tamm_string_array uniq_ind;
            for (auto i: uind) uniq_ind.push_back(strdup(i));
            return uniq_ind;
        }

        default: {
            std::cerr << "Not a valid Expression!\n";
            std::exit(EXIT_FAILURE);
        }
    }
}


//tamm_string_array collectArrayRefs(Exp exp) {
//    ExpList* el = nullptr;
//    tamm_string_array p = nullptr;
//    int c = 0, i = 0;
//    switch (exp->kind) {
//        case is_Parenth:
//            return collectArrayRefs(exp->u.Parenth.exp);
//            break;
//        case is_NumConst:
//            return nullptr;
//            break;
//        case is_ArrayRef:
//            p = tce_malloc(sizeof(*p));
//            p->list = tce_malloc(sizeof(string) * 1);
//            p->list[0] = exp->u.Array.name;
//            p->length = 1;
//            return p;
//            break;
//        case is_Addition:
//            c = 0, i = 0;
//            el = (exp->u.Addition.subexps);
//            string allrefs[65535];
//            while (el != nullptr) {
//                tamm_string_array cref = collectArrayRefs(el->head);
//                for (i = 0; i < cref->length; i++) {
//                    allrefs[c] = cref->list[i];
//                    c++;
//                }
//                el = el->tail;
//            }
//            i = 0;
//            p = tce_malloc(sizeof(*p));
//            p->list = tce_malloc(sizeof(string) * c);
//            p->length = c;
//            for (i = 0; i < c; i++) {
//                p->list[i] = allrefs[i];
//            }
//            return p;
//            break;
//        case is_Multiplication:
//            c = 0, i = 0;
//            el = (exp->u.Multiplication.subexps);
//            string allrefs1[65535];
//            while (el != nullptr) {
//                tamm_string_array cref = collectArrayRefs(el->head);
//                if (cref != nullptr) {
//                    for (i = 0; i < cref->length; i++) {
//                        allrefs1[c] = cref->list[i];
//                        c++;
//                    }
//                }
//                el = el->tail;
//            }
//            i = 0;
//            p = tce_malloc(sizeof(*p));
//            p->list = tce_malloc(sizeof(string) * c);
//            p->length = c;
//            for (i = 0; i < c; i++) {
//                p->list[i] = allrefs1[i];
//            }
//            return p;
//            break;
//        default:
//            std::cerr << "Not a valid Expression!\n";
//            std::exit(EXIT_FAILURE);
//    }
//}
//

File: OLD_TAMM/opmin/src/frontend/semant.py
from error import InputError, FrontEndError
from absyn import Decl, RangeDecl, IndexDecl, ArrayDecl, ExpandDecl, VolatileDecl, IterationDecl, \
     Stmt, AssignStmt, Parenth, NumConst, Array, Addition, Multiplication
from absyn_lib import getIndices
import ast.absyn


def semantCheck(trans_unit):
    for ce in trans_unit.comp_elems:
        __checkCompElem(ce, __SymbolTable())


def __checkCompElem(comp_elem, symtab):
    for e in comp_elem.elems:   # e = each line of the .eq
        __checkElem(e, symtab)


def __checkElem(e, symtab):
    if (isinstance(e, Decl)):   # e is a declaration, i.e. array c_om[O][M]
        __checkDecl(e, symtab)
    elif (isinstance(e, Stmt)):  # e is a statement: hbar[p3,p4,h1,h2] = (-0.5 * t_vo[p3,h3] * v_ovoo[h3,p4,h1,h2])
        __checkStmt(e, symtab)
    else:
        raise FrontEndError('%s: unknown element' % __name__)


def __checkDecl(d, symtab):
    if (isinstance(d, RangeDecl)):      # range O = 10.0;
        __checkRangeDecl(d, symtab)
    elif (isinstance(d, IndexDecl)):    # index h1 = O;
        __checkIndexDecl(d, symtab)
    elif (isinstance(d, ArrayDecl)):    # array f_ov([O][V]);
        __checkArrayDecl(d, symtab)
    elif (isinstance(d, ExpandDecl)):
        __checkExpandDecl(d, symtab)
    elif (isinstance(d, VolatileDecl)):
        __checkVolatileDecl(d, symtab)
    elif (isinstance(d, IterationDecl)):
        __checkIterationDecl(d, symtab)
    else:
        raise FrontEndError('%s: unknown declaration' % __name__)


def __checkRangeDecl(d, symtab):
    __verifyVarDecl(d.name, __getLineNo(d), symtab)
    if ((d.value.value % 1 != 0) or (d.value.value <= 0)):
        raise InputError('"%s" is not a positive integer' % d.value, __getLineNo(d.value))
    symtab.range_tab[d.name] = d.value.value


def __checkIndexDecl(d, symtab):
    __verifyVarDecl(d.name, __getLineNo(d), symtab)
    __verifyRangeRef(d.range.name, __getLineNo(d.range), symtab)
    symtab.index_tab[d.name] = d.range.name


def __checkArrayDecl(d, symtab):
    __verifyVarDecl(d.name, __getLineNo(d), symtab)
    for r in (d.upper_ranges + d.lower_ranges):
        __verifyRangeRef(r.name, __getLineNo(r), symtab)
    __verifySymGroups(d, symtab)
    symtab.array_tab[d.name] = ([r.name for r in d.upper_ranges], [r.name for r in d.lower_ranges])


def __checkExpandDecl(d, symtab):
    if (symtab.expand_tab.has_key(d.arr.name)):
        raise InputError('"%s" already exists in expansion list' % d.arr.name, __getLineNo(d.arr))
    __verifyArrayRefName(d.arr.name, __getLineNo(d.arr), symtab)
    symtab.expand_tab[d.arr.name] = d.arr.name


def __checkVolatileDecl(d, symtab):
    if (symtab.volatile_tab.has_key(d.arr.name)):
        raise InputError('"%s" already exists in volatile list' % d.arr.name, __getLineNo(d.arr))
    __verifyArrayRefName(d.arr.name, __getLineNo(d.arr), symtab)
    symtab.volatile_tab[d.arr.name] = d.arr.name


def __checkIterationDecl(d, symtab):
    if symtab.iteration is not None:
        raise InputError('iteration count is already defined', __getLineNo(d))
    if ((d.value.value % 1 != 0) or (d.value.value <= 0)):
        raise InputError('"%s" is not a positive integer' % d.value, __getLineNo(d.value))
    symtab.iteration = d.value.value


def __checkStmt(s, symtab):
    if (isinstance(s, AssignStmt)):
        __checkAssignStmt(s, symtab)
    else:
        raise FrontEndError('%s: unknown statement' % __name__)


def __checkAssignStmt(s, symtab):
    __checkExp(s.lhs, symtab)
    __checkExp(s.rhs, symtab)
    if (not isinstance(s.lhs, Array)):
        raise InputError('LHS of assignment must be an array reference', __getLineNo(s.lhs))
    elif (s.lhs.isNegative()):
        raise InputError('LHS array cannot be negative', __getLineNo(s.lhs))
    #print str(set(map(lambda x: x.name, getIndices(s.lhs)))) + "=" + str(set(map(lambda x: x.name, getIndices(s.rhs))))
    
    if (set(map(lambda x: x.name, getIndices(s.lhs))) != set(map(lambda x: x.name, getIndices(s.rhs)))):
        raise InputError('LHS and RHS of assignment must have equal index sets', __getLineNo(s))

    lhs_aref = list(__collectArrayRefs(s.lhs))[0]
    rhs_arefs = list(__collectArrayRefs(s.rhs))
    if (lhs_aref in rhs_arefs):
        raise InputError('array "%s" cannot be assigned after being previously referenced (line %s)' %
                         (lhs_aref, __getLineNo(s)), __getLineNo(s))
    for (cur_lhs_aref, cur_rhs_arefs, cur_line_no) in symtab.assign_seq:
        if (lhs_aref in cur_rhs_arefs):
            raise InputError('array "%s" cannot be assigned after being previously referenced (line %s)' %
                             (lhs_aref, cur_line_no), __getLineNo(s))
    symtab.assign_seq.append((lhs_aref, rhs_arefs, __getLineNo(s)))


def __checkExp(e, symtab):
    if (isinstance(e, Parenth)):
        __checkParenth(e, symtab)
    elif (isinstance(e, NumConst)):
        __checkNumConst(e, symtab)
    elif (isinstance(e, Array)):
        __checkArray(e, symtab)
    elif (isinstance(e, Addition)):
        __checkAddition(e, symtab)
    elif (isinstance(e, Multiplication)):
        __checkMultiplication(e, symtab)
    else:
        raise FrontEndError('%s: unknown expression' % __name__)


def __checkParenth(e, symtab):
    __checkExp(e.exp, symtab)


def __checkNumConst(e, symtab):
    pass


def __checkArray(e, symtab):
    __verifyArrayRef(e.name, __getLineNo(e), e.inds, symtab)
    inames = map(lambda x: x.name, getIndices(e))
    rnames = map(lambda x: symtab.index_tab[x], inames)
    (up_range_names, lo_range_names) = symtab.array_tab[e.name]
    istruct = up_range_names + lo_range_names
    if (istruct != rnames):
        raise InputError(
            'array reference "%s%s" must have index structure of "%s%s"' % (e.name, e.inds, e.name, istruct),
            __getLineNo(e)
        )
    for i in inames:
        if (inames.count(i) > 1):
            raise InputError(
                'indistinct index "%s" must not exist in array reference "%s%s"' % (i, e.name, e.inds),
                __getLineNo(e)
            )


def __checkAddition(e, symtab):
    for se in e.subexps:
        __checkExp(se, symtab)
    add_iname_set = set(map(lambda x: x.name, getIndices(e)))
    for se in e.subexps:
        op_iname_set = set(map(lambda x: x.name, getIndices(se)))
        if (add_iname_set != op_iname_set):
            raise InputError('subexpressions of an addition must have equal index sets', __getLineNo(e))


def __checkMultiplication(e, symtab):
    for se in e.subexps:
        __checkExp(se, symtab)
    inames = reduce(lambda x, y: x + y, [map(lambda x: x.name, getIndices(se)) for se in e.subexps], [])
    for i in inames:
        if (inames.count(i) > 2):
            raise InputError('summation index "%s" must occur exactly twice in a multiplication' % i, __getLineNo(e))


def __verifyVarDecl(name, line_no, symtab):
    if (symtab.range_tab.has_key(name) or symtab.index_tab.has_key(name) or symtab.array_tab.has_key(name)):
        raise InputError('"%s" is already defined' % name, line_no)


def __verifyRangeRef(name, line_no, symtab):
    if (not symtab.range_tab.has_key(name)):
        if (symtab.index_tab.has_key(name) or symtab.array_tab.has_key(name)):
            raise InputError('"%s" is not a range' % name, line_no)
        else:
            raise InputError('range "%s" is undefined' % name, line_no)


def __verifyIndexRef(name, line_no, symtab):
    if (not symtab.index_tab.has_key(name)):
        if (symtab.range_tab.has_key(name) or symtab.array_tab.has_key(name)):
            raise InputError('"%s" is not an index' % name, line_no)
        else:
            raise InputError('index "%s" is undefined' % name, line_no)


def __verifyArrayRef(name, line_no, inds, symtab):
    __verifyArrayRefName(name, line_no, symtab)
    for i in inds:
        __verifyIndexRef(i.name, __getLineNo(i), symtab)


def __verifyArrayRefName(name, line_no, symtab):
    if (not symtab.array_tab.has_key(name)):
        if (symtab.range_tab.has_key(name) or symtab.index_tab.has_key(name)):
            raise InputError('"%s" is not an array' % name, line_no)
        else:
            raise InputError('array "%s" is undefined' % name, line_no)


def __verifySymGroups(arr_decl, symtab):
    ranges = arr_decl.upper_ranges + arr_decl.lower_ranges
    for g in arr_decl.sym_groups:
        for ipos in g:
            __checkNumConst(ipos, symtab)
            if ((ipos.value % 1 != 0) or (ipos.value < 0)):
                raise InputError('"%s" is not a valid index position' % ipos.value, __getLineNo(ipos))
            if (ipos.value >= len(ranges)):
                raise InputError('index "%s" is out of range' % int(ipos.value), __getLineNo(ipos))
            all_indexes = map(lambda x: x.value, reduce(lambda x, y: x+y, arr_decl.sym_groups, []))
            if (all_indexes.count(ipos.value) > 1):
                raise InputError('index "%s" cannot occur more than once' % int(ipos.value), __getLineNo(ipos))
        sgroups_ranges = map(lambda x: ranges[int(x.value)].name, g)
        for r in sgroups_ranges:
            pass


def __collectArrayRefs(e):
    if (isinstance(e, Parenth)):
        return __collectArrayRefs(e.exp)
    elif (isinstance(e, NumConst)):
        return set([])
    elif (isinstance(e, Array)):
        return set([e.name])
    elif (isinstance(e, Addition)):
        return reduce(lambda x, y: x | y, [__collectArrayRefs(se) for se in e.subexps], set([]))
    elif (isinstance(e, Multiplication)):
        return reduce(lambda x, y: x | y, [__collectArrayRefs(se) for se in e.subexps], set([]))
    else:
        raise FrontEndError('%s: unknown expression' % __name__)


def __getLineNo(e):
    if (hasattr(e, 'line_no')):
        return e.line_no
    else:
        return None


class __SymbolTable:

    def __init__(self):
        self.range_tab = {}
        self.index_tab = {}
        self.array_tab = {}
        self.expand_tab = {}
        self.volatile_tab = {}
        self.assign_seq = []
        self.iteration = None


def symCheck(trans_unit):
    for ce in trans_unit.comp_elems:
        __symCheckCompElem(ce)


def __symCheckCompElem(comp_elem):
    for e in comp_elem.elems:
        if (isinstance(e, ast.absyn.AssignStmt)):
            __symCheckAssignStmt(e)


def __symCheckAssignStmt(s):
    global_sym_groups = s.lhs.sym_groups
    __symCheckExp(s.rhs, global_sym_groups)


def __symCheckExp(e, global_sym_groups):
    if (isinstance(e, ast.absyn.Array)):
        __symCheckArray(e, global_sym_groups)
    elif (isinstance(e, ast.absyn.Addition)):
        __symCheckAddition(e, global_sym_groups)
    elif (isinstance(e, ast.absyn.Multiplication)):
        __symCheckMultiplication(e, global_sym_groups)
    else:
        raise FrontEndError('%s: unknown declaration' % __name__)


def __symCheckArray(e, global_sym_groups):
    global_sets = map(set, global_sym_groups)
    this_sets = map(set, e.sym_groups)
    for s1 in global_sets:
        found = False
        for s2 in this_sets:
            if (s1 == s2):
                found = True
        if (not found):
            raise InputError(
                'contains an assignment where the permutation symmetry of array "%s" mismatches with LHS' % e.name
            )


def __symCheckAddition(e, global_sym_groups):
    pass


def __symCheckMultiplication(e, global_sym_groups):
    global_sets = map(set, global_sym_groups)
    this_sets = map(set, e.sym_groups)
    for s1 in global_sets:
        for s2 in this_sets:
            if (not ((s1 >= s2) or len(s1 & s2) == 0)):
                raise InputError('contains an assignment where permutation symmetry of LHS is not equal ' +
                                 'or not in higher order than RHS')
